### PR TITLE
determine nullability and type from constructor instead class member for serdes generator

### PR DESCRIPTION
**Pull request by mateominato**

### CODE REVIEW

File: packages/brick_build/lib/src/serdes_generator.dart [hunk 1]
```diff
@@ -196,6 +197,15 @@ abstract class SerdesGenerator<_FieldAnnotation extends FieldSerializable,
   /// If including a custom checker in your domain, overwrite this field
   SharedChecker checkerForType(DartType type) => SharedChecker(type);
 
+  /// Return a `SharedChecker` for a type via field.
```
Comment (tshedor) [inline packages/brick_build/lib/src/serdes_generator.dart:200]: This should be more descriptive. Why is this necessary? Does it relate to other methods? What's the use case? The naming is obvious as to what it does.

File: packages/brick_rest_generators/lib/src/rest_deserialize.dart [hunk 2]
```diff
@@ -46,7 +46,7 @@ class RestDeserialize extends RestSerdesGenerator {
 
       // bool, double, int, num, String
     } else if (checker.isDartCoreType) {
-      return '$fieldValue as ${field.type}$defaultValue';
+      return '$fieldValue as ${checker.targetType}$defaultValue';
```
Comment (tshedor) [inline packages/brick_rest_generators/lib/src/rest_deserialize.dart:49]: There should be test coverage around this too in the `brick_rest_generators` package

File: packages/brick_sqlite_generators/test/sqlite_schema/test_nullable.dart [hunk 3]
```diff
@@ -15,7 +15,9 @@ final schema = Schema(0, generatorVersion: 1, tables: <SchemaTable>{
   SchemaTable('Nullable', columns: <SchemaColumn>{
     SchemaColumn('_brick_id', Column.integer,
         autoincrement: true, nullable: false, isPrimaryKey: true),
-    SchemaColumn('name', Column.varchar, nullable: false)
+    SchemaColumn('name', Column.varchar, nullable: false),
+    SchemaColumn('constructor_field_nullability_mismatch', Column.varchar),
+    SchemaColumn('constructor_field_type_mismatch', Column.varchar)
```
Comment (tshedor) [inline packages/brick_sqlite_generators/test/sqlite_schema/test_nullable.dart:20]: This is good coverage

File: packages/brick_sqlite_generators/lib/src/sqlite_deserialize.dart [hunk 4]
```diff
@@ -65,7 +65,7 @@ class SqliteDeserialize<_Model extends SqliteModel> extends SqliteSerdesGenerato
 
       // double, int, String
     } else if (checker.isDartCoreType) {
-      return '$fieldValue as ${field.type}$defaultValue';
+      return '$fieldValue as ${checker.targetType}$defaultValue';
```
Comment (tshedor) [inline packages/brick_sqlite_generators/lib/src/sqlite_deserialize.dart:68]: There should also be test coverage around this in the `brick_sqlite_generators` package. It's not guaranteed that only OFWIRB will implement this generator

File: packages/brick_offline_first_with_rest_build/test/offline_first_generator/test_nullable_field.dart [hunk 5]
```diff
@@ -7,12 +7,14 @@ Future<NullableField> _$NullableFieldFromRest(Map<String, dynamic> data,
     OfflineFirstRepository? repository}) async {
   return NullableField(
       restFalse: data['rest_false'] as String?,
-      nullableRestTrue: data['nullable_rest_true'] == null
-          ? null
-          : data['nullable_rest_true'] as String?,
+      nullableRestTrue: data['nullable_rest_true'] as String?,
```
Comment (tshedor) [inline packages/brick_offline_first_with_rest_build/test/offline_first_generator/test_nullable_field.dart:10]: There's additional testing and implementation in this package around OfflineFirstSerdes.

**Review by tshedor — CHANGES_REQUESTED** [review]

File: packages/brick_offline_first_with_rest_build/test/offline_first_generator/test_nullable_field.dart (hunk 5 above)
Comment (mateominato) [reply packages/brick_offline_first_with_rest_build/test/offline_first_generator/test_nullable_field.dart:10]: is this actionable? not sure I understand

File: packages/brick_sqlite_generators/lib/src/sqlite_deserialize.dart (hunk 4 above)
Comment (tshedor) [reply packages/brick_sqlite_generators/lib/src/sqlite_deserialize.dart:68]: That's the schema. It doesn't cover the adapter generators

File: packages/brick_offline_first_with_rest_build/test/offline_first_generator/test_nullable_field.dart (hunk 5 above)
Comment (tshedor) [reply packages/brick_offline_first_with_rest_build/test/offline_first_generator/test_nullable_field.dart:10]: The offline first with rest build introduces a new model type, `OfflineFirstSerdes`. The nullable checks around an OfflineFirstSerdes should be tested too, as they aren't as simple as an `as ?`. They use a constructor that cannot accept a nullable object (i.e. `OfflineFirstSerdes.fromRest(data['my_object'])`)

File: packages/brick_build/lib/src/serdes_generator.dart [hunk 6]
```diff
@@ -196,6 +197,27 @@ abstract class SerdesGenerator<_FieldAnnotation extends FieldSerializable,
...
+  /// in a class member definition might not match that of the constructor. In this instance,
+  /// we want to infere type from the constructor, not the field. This requires the class member
+  /// name to match the parameter name in the constructor.
+  ///
+  /// Ex:
+  /// class MyClass {
+  /// final String field;
+  ///
+  /// MyClass({String? field}): field = field ?? 'default';
+  /// }
```
Comment (tshedor) [inline packages/brick_build/lib/src/serdes_generator.dart:212]:
```suggestion
  /// Ex:
  /// ```dart
  ///   class MyClass {
  ///     final String field;
  ///
  ///     MyClass({String? field}): field = field ?? 'default';
  ///   }
  /// ```
```

File: packages/brick_rest_generators/test/rest_model_serdes_generator_test.dart [hunk 7]
```diff
@@ -30,6 +32,11 @@ void main() {
         await generateExpectation(
             'unserializable_field_with_generator', unserializable_field_with_generator.output);
       });
+
+      test('constructor_member_field_mismatch', () async {
```
Comment (tshedor) [inline packages/brick_rest_generators/test/rest_model_serdes_generator_test.dart:36]: it's easier to discover test runs by sync'ing the class name and the test name
```suggestion
      test('RestConstructorMemberFieldMismatch', () async {
```

File: packages/brick_sqlite_generators/lib/src/sqlite_serdes_generator.dart [hunk 8]
```diff
@@ -23,6 +23,16 @@ abstract class SqliteSerdesGenerator<_Model extends SqliteModel>
     required this.repositoryName,
   }) : super(element, fields);
 
+  /// External models, such as REST or I/O, pull type inference data from class constructors.
+  /// This is to allow for class constructors to process potentially nullable data
+  /// into non-nullable data by assigning a default at initialization, or to allow a class
+  /// to process one type into another at construction (eg String myString -> Foo(myString)).
+  /// Sqlite data, however, is always serialized from dart -- as a result, we don't care about
+  /// mismatches between the constructor and member fields, and will always use the type defined
+  /// by the class member.
```
Comment (tshedor) [inline packages/brick_sqlite_generators/lib/src/sqlite_serdes_generator.dart:32]: This is a good sentence.

**Review by tshedor — COMMENTED** [review]
slight change

File: packages/brick_build/example/file_deserialize.dart [hunk 9]
```diff
@@ -25,7 +25,9 @@ class FileDeserialize extends FileSerdesGenerator {
 
       // bool, double, int, num, String
     } else if (checker.isDartCoreType) {
-      return '$fieldValue as ${field.type}$defaultValue';
+      final wrappedCheckerType =
+          wrappedInFuture ? 'Future<${checker.targetType}>' : '${checker.targetType}';
```
Comment (tshedor) [inline packages/brick_build/example/file_deserialize.dart:29]:
```suggestion
          wrappedInFuture ? 'Future<${checker.targetType}>' : checker.targetType.toString();
```

File: packages/brick_build/test/__helpers__.dart [hunk 10]
```diff
@@ -96,7 +96,9 @@ class CustomSerdes extends SerdesGenerator<FieldAnnotation, Model> {
   @override
   String coderForField(field, checker, {required fieldAnnotation, required wrappedInFuture}) {
     final fieldValue = serdesValueForField(field, fieldAnnotation.name, checker: checker);
-    return '$fieldValue as ${field.type}';
+    final wrappedCheckerType =
+        wrappedInFuture ? 'Future<${checker.targetType}>' : '${checker.targetType}';
```
Comment (tshedor) [inline packages/brick_build/test/__helpers__.dart:100]:
```suggestion
        wrappedInFuture ? 'Future<${checker.targetType}>' : checker.targetType.toString();
```

File: packages/brick_rest_generators/lib/src/rest_deserialize.dart [hunk 11]
```diff
@@ -46,7 +47,9 @@ class RestDeserialize extends RestSerdesGenerator {
 
       // bool, double, int, num, String
     } else if (checker.isDartCoreType) {
-      return '$fieldValue as ${field.type}$defaultValue';
+      final wrappedCheckerType =
+          wrappedInFuture ? 'Future<${checker.targetType}>' : '${checker.targetType}';
```
Comment (tshedor) [inline packages/brick_rest_generators/lib/src/rest_deserialize.dart:51]:
```suggestion
          wrappedInFuture ? 'Future<${checker.targetType}>' : checker.targetType.toString();
```

**Review by tshedor — CHANGES_REQUESTED** [review]
Small things but now here's the fun part. 

* Add a CHANGELOG to `brick_build` describing the new methods and preference of using `checkerForField`
* Bump `brick_build`'s version in the pubspec; I'll do a release
* Add a CHANGELOG to `brick_rest_genereators`
* Bump `brick_rest_generators` version and the minimum `brick_build` version
* Add a CHANGELOG to `brick_offline_first_with_rest_build`
* Bump the version and the rest and build version

The non-nullable futures will be a breaking change that should be noted in OFWRB

**Review by tshedor — APPROVED** [review]

**Pull request merged**